### DISCUSSION TITLE
Runtime error in cleanup of magickwand object

### DISCUSSION THREAD

**Discussion by MaheshWalde** [ImageMagick/ImageMagick#6515] [Help]
Hi ,
I am getting following runtime error while using the code below , it converts many gifs, and other types to webp but it crashes during cleanup of those magickWand objects. 

Any idea why its hitting the assertion ?

Error:
 MagickCore/semaphore.c:295: LockSemaphoreInfo: Assertion `semaphore_info != (SemaphoreInfo *) NULL' failed.

My code:

    void cleanupImageMagick(MagickWand *&inputImage, MagickWand *&snapshot)
    {
        if (snapshot) {
            DestroyMagickWand(snapshot);
            snapshot = nullptr;
        }

        if (inputImage) {
            DestroyMagickWand(inputImage);
            inputImage = nullptr;
        }

        MagickWandTerminus();
    }

    //This function converts bmp,tiff,ico, gif images to webp image using ImageMagick library
    bool ConvertOtherTypes(std::string srcImagePath, std::string& destImagePath, std::string mimeType)
    {
        LOG(WARNING)<<"*************** START OTHER_TYPES TO WEBP CONVERSION, MIME TYPE = "<< mimeType<< " **************";
        destImagePath = srcImagePath + ".webp";
        MagickWand *inputImage = NULL , *snapshot = NULL;
        MagickBooleanType status;
        MagickWandGenesis();

        inputImage = NewMagickWand();
        if(!inputImage)
        {
            //TODO add cleanup code in all the failing cases
            LOG(WARNING)<<"UNABLE TO CREATE NEW INPUT IMAGE ";
            cleanupImageMagick(inputImage, snapshot);
            return false;
        }
        if(MagickReadImage(inputImage, srcImagePath.c_str())==MagickFalse)
        {
            LOG(WARNING)<<"FAILED TO READ INPUT IMAGE";
            cleanupImageMagick(inputImage, snapshot);
            return false;
        }

        if(mimeType == "image/gif")
        {
            // Coalesce the frames
            snapshot = MagickCoalesceImages(inputImage);
        }
        else
        {
            snapshot = CloneMagickWand(inputImage);
        }
        if(!snapshot)
        {
            LOG(WARNING)<<"UNABLE TO CREATE CLONE/COALESCE IMAGE ";
            cleanupImageMagick(inputImage, snapshot);
            return false;
        }
        if(mimeType == "image/gif")
        {
            // Set the coalesced images as the active image sequence
            MagickSetIteratorIndex(snapshot, 0);
            // Convert the coalesced frames to WebP
            MagickSetImageFormat(snapshot, "WEBP");
            status = MagickWriteImages(snapshot, destImagePath.c_str(), MagickTrue);
        }
        else
        {
            status = MagickWriteImage(snapshot, destImagePath.c_str());
        }
        if(status==MagickFalse)
        {
            LOG(WARNING)<<"FAILED TO WRITE CONVERTED IMAGE";
            cleanupImageMagick(inputImage, snapshot);
            return false;
        }
        cleanupImageMagick(inputImage, snapshot);
        return true;
    }

**Comment by dlemstra** [comment]
I would advise you to attach a debugger and figure out which call is causing this so you have at lease a call stack. I suspect it has to do with you possibly calling `MagickWandGenesis` and `MagickWandTerminus` multiple times. You should only call those methods at the start and end of your application.

**Reply by MaheshWalde** [reply]
Thanks @dlemstra , 
That was the issue only , I was calling MagickWandGenesis and MagickWandTerminus multiple times. 
Now I am calling them at the start and end only , it solved the issue.
Thank you very much !!!

**Comment by MaheshWalde** [comment]
Need to call _MagickWandGenesis_ and _MagickWandTerminus_ once only at the start and end of application.
